### PR TITLE
Added replace imp id to stored bid responses

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
Implementation of "replaceImpid" from issue https://github.com/prebid/prebid-server/issues/2210

### CODE REVIEW

**Review by bsardo — COMMENTED** [review]
Looks good, just some formatting nitpicks.
Were you planning to handle the "Debug Output" portion of the corresponding issue in a separate PR?

File: exchange/utils.go [hunk 1]
```diff
@@ -760,7 +760,8 @@ func applyFPD(fpd *firstpartydata.ResolvedFirstPartyData, bidReq *openrtb2.BidRe
 
 func buildBidResponseRequest(req *openrtb2.BidRequest,
 	bidderImpResponses stored_responses.BidderImpsWithBidResponses,
-	aliases map[string]string) map[openrtb_ext.BidderName]BidderRequest {
+	aliases map[string]string,
```
Comment (guscarreon) [inline exchange/utils.go:763]: I realize that the code discussed in this comment wasn't added in your PR, but I believe we could address it in order to simplify both this function and `cleanOpenRTBRequests()`. The return value of `stored_responses.InitStoredBidResponses(req.BidRequest, auctionReq.StoredBidResponses)` seems to not be used inside `cleanOpenRTBRequests()` until line 76 in the `buildBidResponseRequest()` call and not used again.  In an effort to make `cleanOpenRTBRequests()` less verbose, can we move the call inside `buildBidResponseRequest()` instead?
```
 40   func cleanOpenRTBRequests(ctx context.Context,
 41       auctionReq AuctionRequest,
 42       requestExt *openrtb_ext.ExtRequest,
 43       bidderToSyncerKey map[string]string,
 44       metricsEngine metrics.MetricsEngine,
 45       gdprDefaultValue gdpr.Signal,
 46       privacyConfig config.Privacy,
 47       gdprPermsBuilder gdpr.PermissionsBuilder,
 48       tcf2ConfigBuilder gdpr.TCF2ConfigBuilder,
 49       hostSChainNode *openrtb2.SupplyChainNode,
 50   ) (allowedBidderRequests []BidderRequest, privacyLabels metrics.PrivacyLabels, errs []error) {
 51  
 52       req := auctionReq.BidRequestWrapper
 53       aliases, errs := parseAliases(req.BidRequest)
 54       if len(errs) > 0 {
 55           return
 56       }
 57  
 58       allowedBidderRequests = make([]BidderRequest, 0, 0)
 59  
 60 -     bidderImpWithBidResp := stored_responses.InitStoredBidResponses(req.BidRequest, auctionReq.StoredBidResponses)
 61  
 62       impsByBidder, err := splitImps(req.BidRequest.Imp)
 63       if err != nil {
 64           errs = []error{err}
 65           return
 66       }
 67  
 68       aliasesGVLIDs, errs := parseAliasesGVLIDs(req.BidRequest)
 69       if len(errs) > 0 {
 70           return
 71       }
 72  
 73       var allBidderRequests []BidderRequest
 74       allBidderRequests, errs = getAuctionBidderRequests(auctionReq, requestExt, bidderToSyncerKey, impsByBidder, aliases, hostSChainNode)
 75  
 76 -     bidderNameToBidderReq := buildBidResponseRequest(req.BidRequest, bidderImpWithBidResp, aliases, auctionReq.BidderImpReplaceImp)
    +     bidderNameToBidderReq := buildBidResponseRequest(req.BidRequest, aliases, auctionReq.BidderImpReplaceImp)
 77       //this function should be executed after getAuctionBidderRequests
 78       allBidderRequests = mergeBidderRequests(allBidderRequests, bidderNameToBidderReq)
 78     allBidderRequests = mergeBidderRequests(allBidderRequests, bidderNameToBidderReq)
 79
 80 *--677 lines: gdprSignal, err := extractGDPR(req.BidRequest)---------------------------------------------------------------------------------------------------------------------------------------------------
757           bidReq.User = fpd.User
758       }
759   }
760
761   func buildBidResponseRequest(req *openrtb2.BidRequest,
762 -     bidderImpResponses stored_responses.BidderImpsWithBidResponses,
763       aliases map[string]string,
764       bidderImpReplaceImpId stored_responses.BidderImpReplaceImpId) map[openrtb_ext.BidderName]BidderRequest {
    +     
    +     bidderImpWithBidResp := stored_responses.InitStoredBidResponses(req.BidRequest, auctionReq.StoredBidResponses)
    +     
765       bidderToBidderResponse := make(map[openrtb_ext.BidderName]BidderRequest)
766 -     for bidderName, impResps := range bidderImpResponses {
    +     for bidderName, impResps := range bidderImpWithBidResp {
767           resolvedBidder := resolveBidder(string(bidderName), aliases)
768           bidderToBidderResponse[bidderName] = BidderRequest{
769               BidRequest:            req,
770               BidderCoreName:        resolvedBidder,
771               BidderName:            bidderName,
772               BidderStoredResponses: impResps,
773               ImpReplaceImpId:       bidderImpReplaceImpId[string(resolvedBidder)],
774               BidderLabels:          metrics.AdapterLabels{Adapter: resolvedBidder},
775           }
776       }
777       return bidderToBidderResponse
778   }
exchange/utils.go
```
Do you think addressing this in this PR is worth it?

**Comment by VeronikaSolovei9** [comment]
> #2210

Per discussion with Bret in the issue I'm not sure "debug" message in http request should be modified. Plus we use it to pass imp id to response. Current debug message clearly shows this is a stored bid response.

File: exchange/utils.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply exchange/utils.go:763]: Oh I see what you mean. 
This is done on purpose. Imagine the situation when we have both real impressions and impressions with stored responses in one request. They can belong to different bidders or the same bidder. 
In InitStoredBidResponses function we extract imps with stored responses and remove them from request. Then we execute existing logic to split reminding  imps by bidder and then we combine them all together to bidder request per bidder. 
As a result we come up with `BidderRequest` where we clearly see what imps should go to MakeRequests and in `BidderStoredResponses` we store imps with stored responses.

Hope it makes sense.

File: exchange/bidder.go [hunk 2]
```diff
@@ -280,7 +280,11 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, bidderRequest Bidde
 							reqBody := string(httpInfo.request.Body)
 							re := regexp.MustCompile(ImpIdReqBody)
 							reqBodySplit := re.Split(reqBody, -1)
-							bidResponse.Bids[i].Bid.ImpID = reqBodySplit[1]
+							reqImpId := reqBodySplit[1]
+							// replace impId if "replaceimpid" is true or not specified
+							if bidderRequest.ImpReplaceImpId[reqImpId] {
+								bidResponse.Bids[i].Bid.ImpID = reqImpId
+							}
```
[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply exchange/bidder.go:287]: Yes, it's a required parameter. Request will not pass validation without it.

File: exchange/utils.go [hunk 3]
```diff
@@ -760,7 +760,8 @@ func applyFPD(fpd *firstpartydata.ResolvedFirstPartyData, bidReq *openrtb2.BidRe
...
+	bidderImpReplaceImpID stored_responses.BidderImpReplaceImpID) map[openrtb_ext.BidderName]BidderRequest {
```
Comment (guscarreon) [inline exchange/utils.go:764]: Nitpick: can we add spaces?
```
761   func buildBidResponseRequest(req *openrtb2.BidRequest,
762       bidderImpResponses stored_responses.BidderImpsWithBidResponses,
763       aliases map[string]string,
764       bidderImpReplaceImpID stored_responses.BidderImpReplaceImpID) map[openrtb_ext.BidderName]BidderRequest {
    +
765       bidderToBidderResponse := make(map[openrtb_ext.BidderName]BidderRequest)
    +
766       for bidderName, impResps := range bidderImpResponses {
767           resolvedBidder := resolveBidder(string(bidderName), aliases)
768           bidderToBidderResponse[bidderName] = BidderRequest{
exchange/utils.go
```

File: stored_responses/stored_responses.go [hunk 4]
```diff
@@ -111,26 +128,41 @@ func extractStoredResponsesIds(impInfo []ImpExtPrebidData,
...
 
 		impIdToStoredResp, impBidderToStoredBidResponse := buildStoredResponsesMaps(storedResponses, impBidderToStoredBidResponseId, impIdToRespId)
-		return impIdToStoredResp, impBidderToStoredBidResponse, nil
+		return impIdToStoredResp, impBidderToStoredBidResponse, bidderImpIdReplaceImp, nil
+	}
+	return nil, nil, nil, nil
+}
+
+// flipMap takes map[impID][bidderName]replaceImpId and modifies it to map[bidderName][impId]replaceImpId
+func flipMap(impBidderReplaceImpId ImpBidderReplaceImpID) BidderImpReplaceImpID {
```
Comment (guscarreon) [inline stored_responses/stored_responses.go:155]: Can we add a `TestFlipMap()` test suite in `stored_responses/stored_responses_test.go`?

**Review by guscarreon — COMMENTED** [review]
Looks pretty good. Can we add a JSON test?

File: stored_responses/stored_responses.go (hunk 4 above)
Comment (VeronikaSolovei9) [reply stored_responses/stored_responses.go:155]: Added

**Review by guscarreon — APPROVED** [review]
LGTM

**Review by bsardo — APPROVED** [review]

**Pull request merged**